### PR TITLE
module 3

### PR DESCRIPTION
**Pull request by TrivialJavaBackender**
JPA and Spring boot

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -26,6 +26,8 @@ subprojects {
         implementation('org.springframework.boot:spring-boot-starter-data-jpa:2.5.5')
         implementation('org.springframework.boot:spring-boot-starter-jdbc:2.5.5')
         implementation('org.springframework.boot:spring-boot-starter-validation:2.5.5')
+        implementation group: 'org.threeten', name: 'threetenbp', version: '1.5.1'
```
Comment (BogdanovaDiana) [inline build.gradle]: Please keep the format consistent with other records

Comment (TrivialJavaBackender) [reply build.gradle]: fixed

File: repository/src/main/java/com/epam/esm/config/JdbcRepositoryConfig.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
+
+import javax.sql.DataSource;
+
+@Configuration
+@ComponentScan("com.epam.esm.repository.impl.jpa")
+@Profile("jdbc")
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/config/JdbcRepositoryConfig.java:19]: A rather ambiguous profile name

File: repository/src/main/java/com/epam/esm/entities/GiftCertificate.java [hunk 3]
```diff
@@ -1,33 +1,179 @@
...
+    @EqualsAndHashCode.Exclude
+    @ManyToMany(cascade = {CascadeType.MERGE})
+    @JoinTable(
+            name = "tag_certificate",
+            joinColumns = @JoinColumn(name="certificate_id"),
+            inverseJoinColumns = @JoinColumn(name="tag_id")
+    )
+    private List<Tag> tags;
+
+    public GiftCertificate(Integer id, String name, String description, BigDecimal price, Double duration, LocalDateTime createDate, LocalDateTime lastUpdateDate) {
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/entities/GiftCertificate.java:59]: Are you using all of these overloaded constructors?

File: repository/src/main/java/com/epam/esm/entities/GiftCertificate.java [hunk 4]
```diff
@@ -1,33 +1,179 @@
...
+        this.description = description;
+        this.price = price;
+        this.duration = duration;
+        this.createDate = createDate;
+        this.lastUpdateDate = lastUpdateDate;
+        this.tags = tags;
+    }
+
+    @PrePersist
+    private void onPrePersist() {
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/entities/GiftCertificate.java]: Method names must begin with a verb

File: repository/src/main/java/com/epam/esm/repository/api/TagRepository.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.epam.esm.entities.Tag;
+
+import java.util.Optional;
+
+/**
+ * Repository for order entity. May perform {@link Savable}, {@link Deletable}, {@link Countable} and
+ * {@link Findable} operations.
+ */
+public interface TagRepository extends Findable<Tag>, Deletable<Tag>, Savable<Tag>, Countable {
```
Comment (iyanushkevich) [inline repository/src/main/java/com/epam/esm/repository/api/TagRepository.java:11]: Could you explain the purpose of all these interfaces?

File: service/src/main/java/com/epam/esm/data/PageInfo.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+public class PageInfo {
+
+    private final int pageSize;
+    private final int currentPage;
+    private final int entityCount;
+    private final int lastPage;
+
+    public PageInfo(int pageSize, int currentPage, int entityCount) {
+        this.pageSize = Math.max(pageSize, 1);
```
Comment (iyanushkevich) [inline service/src/main/java/com/epam/esm/data/PageInfo.java]: Let`s use separate variable for this number

File: service/src/main/java/com/epam/esm/data/PageInfo.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    public PageInfo prevOrFirst() {
+        if (hasPrevious()) {
+            return prev();
+        }
+        return this;
+    }
+
+    public PageInfo first() {
```
Comment (iyanushkevich) [inline service/src/main/java/com/epam/esm/data/PageInfo.java]: Method name as well

File: repository/src/main/java/com/epam/esm/repository/api/TagRepository.java (hunk 5 above)
Comment (TrivialJavaBackender) [reply repository/src/main/java/com/epam/esm/repository/api/TagRepository.java:11]: By the requirements different operations on enttites must be performed: Save, get and delete on tag, only get on user and so on. So I segregated entity interfaces on smaller interfaces for more flexible control.

File: repository/src/main/java/com/epam/esm/entities/GiftCertificate.java (hunk 3 above)
Comment (TrivialJavaBackender) [reply repository/src/main/java/com/epam/esm/entities/GiftCertificate.java:59]: No. Deleted

File: repository/src/main/java/com/epam/esm/config/JdbcRepositoryConfig.java (hunk 2 above)
Comment (TrivialJavaBackender) [reply repository/src/main/java/com/epam/esm/config/JdbcRepositoryConfig.java:19]: Fixed wrong package on component scan

File: repository/src/main/java/com/epam/esm/entities/GiftCertificate.java (hunk 4 above)
Comment (TrivialJavaBackender) [reply repository/src/main/java/com/epam/esm/entities/GiftCertificate.java]: fixed

File: service/src/main/java/com/epam/esm/data/PageInfo.java (hunk 7 above)
Comment (TrivialJavaBackender) [reply service/src/main/java/com/epam/esm/data/PageInfo.java]: renamed methods

**Review by iyanushkevich — APPROVED** [review]

**Pull request merged**